File: b1.m
% Define parameters
Tss = 0.35; % single support paretic stance duration in seconds
T1 = 0.15 * Tss; % 15% of Tss
fs = 10e3; % sampling frequency in Hz
N = round(T1 * fs); % number of samples
t = linspace(0, T1, N); % time vector
theta_i = 5; % initial angle in degrees
theta_f = 20; % final angle in degrees
b0 = 3; % initial slot height in degrees
b11 = 1; % final slot height in degrees

% Commanded reference trajectory (straight line)
theta_ref = linspace(theta_i, theta_f, N);

% Calculate top and bottom edge trajectories of the slot
b = linspace(b0, b11, N); % slot height decreases linearly
t_edge = theta_ref + b/2; % top edge of the slot
b_edge = theta_ref - b/2; % bottom edge of the slot

% Plot the slot and reference trajectory
figure;
plot(t, t_edge, 'r', t, b_edge, 'r', t, theta_ref, 'b');
legend('Top edge', 'Bottom edge', 'Reference trajectory', 'Location','northwest');
xlabel('Time (s)');
ylabel('Angle (degrees)');
title('Slot and Reference Trajectory')

% Absolute positions of the top and bottom edges of the slot at 20 ms
i_20ms = round(20e-3 * fs); % index corresponding to 20 ms
t_edge_20ms = t_edge(i_20ms);
b_edge_20ms = b_edge(i_20ms);
fprintf('At 20 ms, the top edge is at %.2f degrees and the bottom edge is at %.2f degrees.\n', t_edge_20ms, b_edge_20ms);

File: B1.py
import numpy as np
import matplotlib.pyplot as plt

# Given values
theta_0 = 5  # initial angle in degrees
theta_1 = 20  # commanded angle in degrees
T_ss = 0.35  # single support paretic stance duration in seconds
b_0 = 3  # initial slot height in degrees
b_1 = 1  # final slot height in degrees
sampling_frequency = 10000  # 10 kHz

# Step 1: Calculate commanded reference trajectory
N = int(T_ss * sampling_frequency)
i_values = np.arange(N)
theta_ref_i = theta_0 + (i_values / N) * (theta_1 - theta_0)

# Step 2: Calculate slot dynamics
t_i = theta_ref_i + b_0
b_i = theta_ref_i - b_1

# Step 3: Plot the dynamics
plt.plot(i_values, t_i, label='Top Edge')
plt.plot(i_values, b_i, label='Bottom Edge')
plt.xlabel('Sampling Instant (i)')
plt.ylabel('Angle (degrees)')
plt.legend()
plt.title('Slot Dynamics')
plt.show()

# Step 4: Check if the slot collapses as desired (visual inspection)

# Step 5: Calculate positions at 20 ms
position_at_20ms_index = int(0.02 * sampling_frequency)
position_at_20ms = t_i[position_at_20ms_index], b_i[position_at_20ms_index]
print(f'Positions at 20 ms: Top Edge = {position_at_20ms[0]} degrees, Bottom Edge = {position_at_20ms[1]} degrees')
